### ISSUE TITLE
Cannot use features + workspace from cargo tarpaulin (works from cargo test)

### ISSUE THREAD

**Issue by djeedai** [xd009642/tarpaulin#631]
Using tarpaulin 0.16.0 due to a bug in later version(s).

I am trying to invoke tarpaulin to build a **single** unified Lcov report (because coveralls.io only supports 1 report per build per GitHub repo) of a workspace with 3 different runs:
- library A with feature set X
- library A with feature set Y
- binary B

Previously the project was configured without workspace (there was no binary B) and everything worked fine via a `tarpaulin.toml` config:

```toml
[X]
no-default-features = true
features = "default"

[Y]
no-default-features = true
features = ""

[report]
out = ["Lcov"]
```

Here feature set X is "default", and feature set Y is no feature at all. This produced a single Lcov report as expected, which merged the results of both runs and gave a global coverage for the project. All good.

Now converting the GitHub repository and adding workspace, and things are falling over. First there is a limitation (from cargo I think) which prevents using `--features` from the root of the (virtual) workspace. This can be worked around in a single build or test by using `--manifest-path`, _e.g._ manually:

```
cd <PROJECT_ROOT>
cargo build --manifest-path lib/Cargo.toml --no-default-features --features "default"
cargo build --manifest-path lib/Cargo.toml --no-default-features --features ""
cargo build --manifest-path bin/Cargo.toml
```

This works with `cargo test` too and will do what is expected : build/test the specific package with the specific feature set, and nothing else.

However when used with tarpaulin, it seems this doesn't work. I tried to specify `manifest-path = "lib"` in each coverage variant:

```toml
[X]
manifest-path = "lib/Cargo.toml"
no-default-features = true
features = "default"

[Y]
manifest-path = "lib/Cargo.toml"
no-default-features = true
features = ""

[B]
manifest-path = "bin/Cargo.toml"
```

but that doesn't work, and cargo still complains that `--features` cannot be used, which is the error one would get manually if **not** specifying `--manifest-path`:

```
error: --features is not allowed in the root of a virtual workspace
note: while this was previously accepted, it didn't actually do anything
```

It seems that `cargo tarpaulin` is not invoking the build/test the same way a direct command-line invoke of `cargo build` or `cargo test` does. I see no way to work around this, and this is blocking me from using a workspace and adding a binary crate to my repository.

**Comment by xd009642** [comment]
I think if you use the `--debug` flag you should be able to see the flags I'm passing into cargo, if you can do that and send them to me that would be helpful. Also, if you can send me an example project I can try to have a look (I'm abroad right now so not able to dedicate as much time to tarpaulin as normal).

**Comment by djeedai** [comment]
Thanks @xd009642, good tip and great project!

I ran with `--debug` and I see that `manifest-path` is wrong. The value set in `tarpaulin.toml` is never used; instead the repository root is passed, which now explains why cargo is complaining about `--features` since the root is a virtual workspace.

That is, given the above TOML config file with `manifest-path = "lib/Cargo.toml"`, the flags passed to cargo is `--manifest-path Cargo.toml` instead (the repo root):

```
$ cargo tarpaulin --verbose --debug --timeout 120 --out Lcov --config tarpaulin.toml
Nov 27 10:29:11.838 DEBUG cargo_tarpaulin: set up logging
Nov 27 10:29:11.838  INFO cargo_tarpaulin::config: Creating config
Nov 27 10:29:11.838 TRACE cargo_tarpaulin: Debug mode activated
Nov 27 10:29:11.862  INFO cargo_tarpaulin: Running config default_coverage
Nov 27 10:29:11.862  INFO cargo_tarpaulin: Running Tarpaulin
Nov 27 10:29:11.862  INFO cargo_tarpaulin: Building project
Nov 27 10:29:11.883 TRACE cargo_tarpaulin::cargo: Setting RUSTDOCFLAGS=' -C relocation-model=dynamic-no-pic -C link-dead-code -C debuginfo=2 --cfg=tarpaulin  --persist-doctests /home/<user>/dev/weldr/target/doctests -Z unstable-options '
Nov 27 10:29:11.883 TRACE cargo_tarpaulin::cargo: Running command "cargo" "+stable-x86_64-unknown-linux-gnu" "test" "--no-run" "--message-format" "json" "--manifest-path" "/home/<user>/dev/weldr/Cargo.toml" "--tests" "-vvv" "--features" "default" "--no-default-features" "--target-dir" "/home/<user>/dev/weldr/target"
error: --features is not allowed in the root of a virtual workspace
note: while this was previously accepted, it didn't actually do anything
```

Indeed the `manifest-path` option is unconditionally overwritten by the command-line argument, even if not provided on the command line, and the value in `tarpaulin.toml` is always ignored:

https://github.com/xd009642/tarpaulin/blob/e3f3f857ef081aa4d356df074389870f84f8fa30/src/config/mod.rs#L472

**Comment by xd009642** [comment]
ah cool it should be checking the default_manifest and preferring a non-default path. If you have a project I can try on I may have fixed the problems in (the yanked) 0.17.0 release. So I can test my fixes on that crate and then also fix this and release a 0.17.1 hopefully soon

**Comment by djeedai** [comment]
My work-in-progress branch of my `weldr` crate exhibit the issue. It's not a minimal example, but is fairly small. Commit `9ee4bb6e5f` has the code configured for using `manifest-path` in the config file:

https://github.com/djeedai/weldr/tree/9ee4bb6e5f72a2d9dada7ccc43732c47460d9f27

You can have a look at the CI script to see how I actually invoke tarpaulin, but it's same as above.

https://github.com/djeedai/weldr/blob/9ee4bb6e5f72a2d9dada7ccc43732c47460d9f27/.github/workflows/ci.yml#L76-L79

**Comment by xd009642** [comment]
@djeedai so I've tried with this https://github.com/xd009642/tarpaulin/pull/632 and your project and I've seen it go from not working to working. Also this has the features from 0.17 I yanked with my recent fixes so it should all work. I'd appreciate if you try it out and let me know. 

I'm going to fill in some more stuff so all relative paths in the tarpaulin.toml are made relative to the location of the config file for completeness and make sure there's no similar config-merge-bungling. Add a test project that recreates the environment and then it should all be good

**Comment by xd009642** [comment]
So test is done and this is merged. It's in develop and works for your project. I'll aim to have a release out in the next week

**Comment by djeedai** [comment]
Thanks @xd009642 for the fix, this is much appreciated! Sorry I didn't get a notification on that mention.

I've actually come to realize doing other works that the (only) feature I was using on my project needed to become a mandatory dependency, so I removed the feature, and the only test I have is the outdated code I gave you. Going forward I'm sure this fix will come handy to me anyway as soon as I re-add a feature for something else, and no doubt others will benefit from it too. But in the meantime I don't have much else to test with. I will however give 0.17 a try on my CI and let you know if that solves the random failures I had.

Can I delete the test branch where I had those features and which exhibited the bug? I was keeping it alive for your testing, but I'd prefer to clean it up if you don't have any more use for it, as the work on it is actually merged to `main` in my project. Let me know, I don't mind keeping it alive if it's useful for your testing.
